doc/release/upcoming_changes/16200.compatibility.rst

```diff
@@ -0,0 +1,64 @@
+NumPy Scalars are cast when assigned to arrays
+----------------------------------------------
+
+When creating or assigning to arrays, in all relevant cases NumPy
+scalars will now be cast identically to NumPy arrays.  In particular
+this changes the behaviour in some cases which previously raised an
+error::
+
+    np.array([np.float64(np.nan)], dtype=np.int64)
+
+will succeed at this time (this may change) and return an undefined result
+(usually the smallest possible integer).  This also affects assignments::
+
+    arr[0] = np.float64(np.nan)
+
+Note, this already happened for ``np.array(np.float64(np.nan), dtype=np.int64)``
+and that the behaviour is unchanged for ``np.nan`` itself which is a Python
+float.
```

Comment:
How hard would it be to make any conversion of `nan` or `inf` to any int raise? I understand there might be benchmark concerns which could be checked with an appropriate benchmark.

Reply:
Well, the question is what we want, do we want NumPy scalars (which are almost like 0-D arrays) behave like arrays, or do we want them to behave specially.  If we do not want special casing, my code is probably nicer.

That would mean adding a range check to the actual float casting code, I did not check how much speed that costs, for large arrays probably none (memory speeds should be much slower)

Reply:
I was thinking that in all cases: arrays, 0-d arrays, and scalars we do not want to convert `nan` and `inf` to ints by default. If someone actually wants this (for arrays and 0-d arrays) they could do it via view -> copy.

Reply:
Well, I don't much like 0-D arrays being special. And Scalars and 0-D arrays are almost the same (because 0-D arrays tend to be converted to scalars). But yes, you could define element setting as being special...

In general, that does *not* work, of course for most of our dtypes it does:

* Integer element assignment calls `__int__`.
* Float element assignment calls `__float__`
* Complex element assignment can use `__complex__`
* String/Bytes array elements might be able to assume `__str__` is fine

And that works mostly. It currently creates bugs, because `complex128` thinks it can reasonably use `__complex__`, which it must not for a `float128`.  It also means you that 0-D objects tend to behave special, because instead of aligning how the two work, we define them as two different operations.  And that *can* be confusing `str(array-like)` is well defined after all, we currently solve that by refuse to call `str(obj)` for any sequence).  Also within `float128`/`complex128` we (incompletely) hack around these by checking for `0-D` arrays.

So yes, I can make single element assignment special (in most cases, I have my doubt that real consistency can be achieved).  But, we can only do that for some of our own dtypes, where Python has well defined protocols.
So either, I need to flag these dtypes to behave differently (let them always just use `__float__`, `__int__`, etc., except for times and float128/complex128).  Or do some other dance, which feels like it makes dtype creation unnecessarily complicated. `DType.setitem` would have to call/signal back into NumPy in some cases.

Just for the sake of it:
```
arr = np.empty(2, dtype=np.int64)
arr[:] = [np.nan, np.nan]  # fails `np.array([np.nan, np.nan], dtype=np.int64)` uses item assignment after all
arr[:] = np.array([np.nan, np.nan])  # succeeds
arr[0] = np.array(np.nan)  # fails
arr[0, ...] = np.array(np.nan)  # succeeds
```
Is the current behaviour, I cannot say it illogical, but I wonder if it is the best solution to cement item assignment as being special and whether it is really easier on users.

But again, yes, I can put these behaviours back in place, they can just *not* generalize to arbitrary user-defined dtypes. And maybe we have to discuss it on wednesday or the mailing list...

Reply:
In my opinion, all of the examples of assignment should fail. Any reasonable user-defined dtype that converts from float to int should also fail when the conversion makes no sense. Hopefully we will not get the blame if that happens.

Reply:
I do not disagree, but in that case *casting* should probably fail in general (or at least warn) and most just the 0-D case, right?
So it is IMO not an argument for adding an extra path for `setitem`, except possibly performance reasons.  The question is whether we want to special case these 0-D item assignments, because the behaviour is *better* or just because it is somewhat *faster* than doing proper casting?

Reply:
From the sidelines: agreed that casting should fail in those cases (or, more generally, that setting should behave identically to casting the value to the dtype of the array being set).

Reply:
So for me right now, this feels like adding a fast-path, since I don't like the difference between the two paths (casting vs. item assignment).  But of course we can argue, that if the long goal is to align casting with current item assignment behaviour, we can add those fast-paths right now.  (It must be a fast/additional paths, since Python protocols are not a generic solution.)

I don't mind adding the fast path now, I feel it is one more thing to think about, but, it does not need to be exposed to user-defined dtypes right away.  They are unlikely to make use of this, except maybe a rational dtype using "as integer ratio".  (I am not 100% sure there cannot be reasons to diverge from the Python scalar protocols.)

I am unsure item assignment+casting is actually worth fast-paths, but maybe that does not matter.  They currently lead to weird things (e.g. a DNA dtype, which is much like a string and a sequence could never be assigned to a NumPy string dtype!).  However, those issues are resolvable now in either case.

Reply:
Is there an open issue or PR to make any conversion from `inf` or `nan` to `int` raise an error or at least warn?

Reply:
https://github.com/numpy/numpy/issues/16624